polkadot/cli/src/lib.rs

```diff
@@ -145,19 +137,27 @@ pub fn run<I, T>(args: I) -> error::Result<()> where
 	info!("  version {}", crate_version!());
 	info!("  by Parity Technologies, 2017, 2018");
 
+	if let Some(matches) = matches.subcommand_matches("build-spec") {
+		let spec = load_spec(&matches);
+		info!("Building chain spec");
+		let json =  if matches.is_present("raw") {
+			spec.to_json_raw()
+		} else {
+			spec.to_json()
+		};
+		print!("{}", json);
+		return Ok(())
+	}
+
 	let mut config = service::Configuration::default();
 
 	if let Some(name) = matches.value_of("name") {
 		config.name = name.into();
 		info!("Node name: {}", config.name);
 	}
 
-	let chain_spec = matches.value_of("chain")
-		.map(ChainSpec::from)
-		.unwrap_or_else(|| if matches.is_present("dev") { ChainSpec::Development } else { ChainSpec::PoC2Testnet });
-	info!("Chain specification: {}", chain_spec);
-
-	config.chain_name = chain_spec.clone().into();
+	let spec = load_spec(&matches);
```

Comment:
This is expensive, isn't it? If it loads a potentially huge raw genesis map or genesis config and builds genesis from it, then it's needless most of the time. Stuff should only be loaded if genesis block is needed to be built as it was before.

Reply:
Before it was loaded unconditionally as well, wasn't? `genesis_storage` is required in `config` currently. I'll add an issue to optimise this later, as it requires partial json parsing.

Reply:
I see now it is deferred in latest master.

Reply:
no - `genesis_storage` was a function that only bothered doing the work if the genesis needed to be computed.

Reply:
would prefer to have the load-on-demand done in this pr, but if it's really going to add significant complexity, then i guess it could be split off for the next pr.

Reply:
Implemented